channel/contractbackend.go

```diff
@@ -62,25 +62,116 @@ type Transactor interface {
 // This is needed to send on-chain transaction to interact with the smart contracts.
 type ContractBackend struct {
 	ContractInterface
-	tr                Transactor
+	tr              Transactor
+	noncer          noncer
+	txFinalityDepth uint64
+	chainID         ChainID
+}
+
+type noncer interface {
+	setNonce(chainID ChainID, addr common.Address, nonce uint64) (uint64, error)
+}
+
+// LocalNoncer implements noncer interface.
+type LocalNoncer struct {
+	expectedNextNonce map[common.Address]uint64
 	nonceMtx          *sync.Mutex
+}
+
+// GlobalNoncer implements a global nonce counter that is shared across different contract backends.
+type GlobalNoncer struct {
 	expectedNextNonce map[common.Address]uint64
-	txFinalityDepth   uint64
-	chainID           ChainID
+	nonceMtx          map[common.Address]*sync.Mutex
+}
```

Comment:
The `GlobalNoncer` requires a mutex to guarantee only a single nonce is created for each `address` entry.
Furthermore I would rename the struct to `SharedNoncer`. But I am honestly a little bit surprised by the definition.
`LocalNoncer` and `GlobalNoncer` effectively do the same: Guarding access to the `Nonce` for a specific `common.Address`. The only difference is, that the `LocalNoncer` is more restrictive in that it only allows updates to one nonce entry at a time.
The `GlobalNoncer` on the other hand allows updates to multiple nonces at the same time within the same map. Thinking about it even more this might create another issue where two routines try to write to the same map at once, which would require another mutex anyway, making `GlobalNoncer` just a more complicated and convoluted implementation of the `LocalNoncer`.

Long story short: I would get rid of the `GlobalNoncer`, rename `LocalNoncer` to `SharedNoncer` and accept the simpler implementation of a single mutex guarding all nonces.